tests/test_streams.py

```diff
@@ -266,3 +288,121 @@ def test_dump_varint_positive(tmp_path):
+        pytest.skip(
+            "Maven compatibility-test.jar build failed (maybe Java version <11?)"
+        )
+
+
+jar = "tests/streams/java/target/compatibility-test.jar"
+
+
+def run_jar(command: str, tmp_path):
+    return run(["java", "-jar", jar, command, tmp_path], check=True)
+
+
+def run_java_single_varint(value: int, tmp_path) -> int:
+    # Write single varint to file
+    with open(tmp_path / "py_single_varint.out", "wb") as stream:
+        betterproto.dump_varint(value, stream)
+
+    # Have Java read this varint and write it back
+    run_jar("single_varint", tmp_path)
+
+    # Read single varint from Java output file
+    with open(tmp_path / "java_single_varint.out", "rb") as stream:
+        returned = betterproto.load_varint(stream)
+        with pytest.raises(EOFError):
+            betterproto.load_varint(stream)
+
+    return returned
+
+
+def test_single_varint(compile_jar, tmp_path):
+    single_byte = (1, b"\x01")
+    multi_byte = (123456789, b"\x95\x9A\xEF\x3A")
+
+    # Write a single-byte varint to a file and have Java read it back
+    returned = run_java_single_varint(single_byte[0], tmp_path)
+    assert returned == single_byte
+
+    # Same for a multi-byte varint
+    returned = run_java_single_varint(multi_byte[0], tmp_path)
+    assert returned == multi_byte
+
+
+def test_multiple_varints(compile_jar, tmp_path):
+    single_byte = (1, b"\x01")
+    multi_byte = (123456789, b"\x95\x9A\xEF\x3A")
+    over32 = (3000000000, b"\x80\xBC\xC1\x96\x0B")
+
+    # Write two varints to the same file
+    with open(tmp_path / "py_multiple_varints.out", "wb") as stream:
+        betterproto.dump_varint(single_byte[0], stream)
+        betterproto.dump_varint(multi_byte[0], stream)
+        betterproto.dump_varint(over32[0], stream)
+
+    # Have Java read these varints and write them back
+    run_jar("multiple_varints", tmp_path)
+
+    # Read varints from Java output file
+    with open(tmp_path / "java_multiple_varints.out", "rb") as stream:
+        returned_single = betterproto.load_varint(stream)
+        returned_multi = betterproto.load_varint(stream)
+        returned_over32 = betterproto.load_varint(stream)
+        with pytest.raises(EOFError):
+            betterproto.load_varint(stream)
+
+    assert returned_single == single_byte
+    assert returned_multi == multi_byte
+    assert returned_over32 == over32
+
+
+def test_single_message(compile_jar, tmp_path):
+    # Write message to file
+    with open(tmp_path / "py_single_message.out", "wb") as stream:
+        oneof_example.dump(stream)
+
+    # Have Java read and return the message
+    run_jar("single_message", tmp_path)
+
+    # Read and check the returned message
+    with open(tmp_path / "java_single_message.out", "rb") as stream:
+        returned = oneof.Test().load(stream, len(bytes(oneof_example)))
+        assert stream.read() == b""
+
+    assert returned == oneof_example
+
+
+def test_multiple_messages(compile_jar, tmp_path):
```

Comment:
This test assumes that both the Java side and the Python side know the number of messages to read (two in this case - `oneof.Test` & `nested.Test`).

I think we also need to test that EOF can be correctly detected when neither the Java side nor the Python side know the number of messages they have to read.

Reply:
I've added a `test_infinite_messages` that I believe does what you're looking for here - let me know if it looks good or if it's not what you were looking for.